theano/sandbox/cuda/nnet.py

```diff
@@ -721,3 +725,120 @@ def c_support_code_apply(self, node, nodename):
         return ret1 + "\n" + ret2
 
 gpu_softmax_with_bias = GpuSoftmaxWithBias()
+
+
+def hierarchical_softmax(W1, b1, W2, b2, x, n_outputs, target=None):
+    """
+    GPU-only function that returns the outputs of a two-level hierarchical
+    softmax.
+    In the two-level hierarchical softmax architecture, outputs are grouped
+    in sqrt(n_outputs) classes.
+    There are two softmax layers. The first predicts the class of the input x
+    while the second predicts the output of the input x in the predicted class.
+    More explanations can be found in the original paper
+    http://arxiv.org/abs/cs/0108006.
+    If target is specified, it will only compute the outputs of the
+    corresponding targets. Otherwise, if target is None, it will compute all
+    the outputs.
+
+    :type W1: a Theano shared variable of shape (number of features of the
```

Comment:
I would describe parameters as [here](https://github.com/Theano/Theano/blob/master/theano/tensor/nnet/conv.py#L39). Shape is defined in ```:param name:``` and ```:type name:``` is simply ```symbolic 2D tensor``` (which holds for both TensorVariable and SharedVariable).